### PR TITLE
[HW] Update new builder to not call `erase` on OutputOp directly.

### PR DESCRIPTION
**Pull request by mikeurbach**
This builder might be used with a PatternRewriter, which requires all IR mutation such as erases to go through the PatternRewriter. Instead, add a flag to the generic builder to decide if an OutputOp terminator should be created. In the new builder, this is set to false, and the correct OutputOp is created directly, rather than erasing the old one and creating a new one.

I tested this on c4e2720, where I was able to reproduce the issue with calling `erase` directly, and this fixed the issue. I also tested this on top-of-tree.

### CODE REVIEW

**Review by mortbopet — APPROVED** [review]
Awesome. Thank you for locating the error and finding a fix - highly appreciated!

**Review by richardxia — APPROVED** [review]
LGTM, thanks! I can confirm that when I cherry-pick this patch onto c4e272064d76e75dab7de7d32c5857e01418b5d8, the CalyxToHW tests no longer fail with ASAN enabled.

**Comment by mikeurbach** [comment]
Thanks for the quick reviews. Onward!

File: lib/Dialect/HW/HWOps.cpp [hunk 1]
```diff
@@ -603,22 +604,21 @@ void HWModuleOp::build(OpBuilder &builder, OperationState &odsState,
                        HWModuleBuilder modBuilder, ArrayAttr parameters,
                        ArrayRef<NamedAttribute> attributes,
                        StringAttr comment) {
-  build(builder, odsState, name, ports, parameters, attributes, comment);
+  build(builder, odsState, name, ports, parameters, attributes, comment,
+        /*shouldEnsureTerminator=*/false);
   auto *bodyRegion = odsState.regions[0].get();
   OpBuilder::InsertionGuard guard(builder);
   builder.setInsertionPointToStart(&bodyRegion->front());
```
Comment (dtzSiFive) [inline lib/Dialect/HW/HWOps.cpp:611]: Not introduced in this change, but why does this set the insertion point, only to set it again right away?
(AFAICT `builder` isn't used and no IR is changed in the constructor of the accessor)

Consider setting it just once.

**Review by dtzSiFive — COMMENTED** [review]
Awesome, thanks for chasing this down and fixing!

Small question re:insertion point, but it's unrelated really so no worries.

File: lib/Dialect/HW/HWOps.cpp (hunk 1 above)
Comment (mikeurbach) [reply lib/Dialect/HW/HWOps.cpp:611]: Good point!

**Pull request merged**